crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs

```diff
@@ -17,84 +18,253 @@ pub(crate) enum PeerDiscoverResult {
+        let url = url?;
+
+        // if they are local, return the shortcut result
+        if inner.i_s.is_agent_local(agent).await? {
+            return Ok(PeerDiscoverResult::OkShortcut);
+        }
+
+        // attempt an outgoing connection
+        let con_hnd = inner.ep_hnd.get_connection(url.clone(), timeout).await?;
+
+        // return the result
+        Ok(PeerDiscoverResult::OkRemote {
+            url: url.into(),
+            con_hnd,
+        })
+    }
+    .map(|r| match r {
+        Ok(r) => r,
+        Err(e) => PeerDiscoverResult::Err(e),
+    })
+}
+
+/// looping search for agents covering basis_loc
+/// by requesting closer agents from remote nodes
+pub(crate) fn search_remotes_covering_basis(
+    inner: Arc<SpaceReadOnlyInner>,
+    basis_loc: u32,
+    timeout: KitsuneTimeout,
+) -> impl Future<Output = KitsuneP2pResult<Vec<AgentInfoSigned>>> + 'static + Send {
+    const INITIAL_DELAY: u64 = 100;
+    const MAX_DELAY: u64 = 1000;
+    const CHECK_NODE_COUNT: usize = 8;
 
+    async move {
+        let backoff = timeout.backoff(INITIAL_DELAY, MAX_DELAY);
         loop {
-            if let Ok(res) = check_local().await {
-                return res;
+            let mut cover_nodes = Vec::new();
+            let mut near_nodes = Vec::new();
+
+            // first check our local peer store,
+            // sort into nodes covering the basis and otherwise
+            for node in get_cached_remotes_near_basis(inner.clone(), basis_loc, timeout)
+                .await
+                .unwrap_or_else(|_| Vec::new())
+            {
+                if node.storage_arc.contains(basis_loc) {
+                    cover_nodes.push(node);
+                } else {
+                    near_nodes.push(node);
+                }
+                if cover_nodes.len() + near_nodes.len() >= CHECK_NODE_COUNT {
+                    break;
+                }
             }
 
-            if let Ok(res) = check_peer_store().await {
-                return res;
+            // if we have any nodes covering the basis, return them
+            if !cover_nodes.is_empty() {
+                return Ok(cover_nodes);
             }
 
-            let elapsed_ms = start_time.elapsed().as_millis() as u64;
-            if elapsed_ms >= timeout_ms {
-                return PeerDiscoverResult::Err("timeout".into());
+            if near_nodes.is_empty() {
+                // maybe just wait and try again?
+                backoff.wait().await;
+                continue;
             }
 
-            interval_ms *= 2;
-            if interval_ms > timeout_ms - elapsed_ms {
-                interval_ms = timeout_ms - elapsed_ms;
+            // the next step involves making network requests
+            // so check our timeout first
+            timeout.ok()?;
+
+            // shuffle the returned nodes so we don't keep hammering the same one
+            use rand::prelude::*;
+            near_nodes.shuffle(&mut rand::thread_rng());
```

Comment:
This rng is happening inside a double nested loop. Might be worth reusing it?

Reply:
No, we cannot cache thread-local data across awaits, they are not Send, because they are thread local... also thread-local is very fast, you don't need any locks around it, because there cannot be contention : )